\input{../header_1_en.tex}

\SweaveOpts{prefix.string=plots/fig,eps=FALSE,echo=FALSE}

<<echo=FALSE, eval=TRUE>>=
options(width=40)
@ 

\begin{document}

\begin{frame}[allowframebreaks]
  \titlepage
\end{frame}

\section*{Class outline}

\begin{frame}[allowframebreaks]
  \frametitle{A Case Study with GOs}
  \tableofcontents[hideallsubsections]
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Intro}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Packages we'll use today}
  \begin{itemize}
  \item You'll probably need to install a few using \BIOCfunction{biocLite}.
<<eval=FALSE, echo=TRUE>>=
library("ALL")
library("Biobase")
library("annotate")
library("hgu95av2.db")
library("genefilter")
library("annaffy")
library("GO.db")
library("GOstats")
library("biomaRt")
library("hgu133a.db")
library("lattice")
@  
  \end{itemize}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data and filtering}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{To start off}
  \begin{itemize}
  \item Similar to the 2nd homework, lets create a subset from the \pl{ALL} dataset.
  \item Remember that we are working with leukemia samples and the molecular types BCR/ABL and ALL/AF4 are different translocations.
<<eval=TRUE, echo=TRUE>>=
library("ALL")
data("ALL")
types <- c("ALL1/AF4", "BCR/ABL")
bcell <- grep("^B", as.character(ALL$BT))
ALL_af4bcr <- ALL[, intersect(bcell,
	which(ALL$mol.biol %in% types))]
ALL_af4bcr$mol.biol <- factor(ALL_af4bcr$mol.biol)
@  
  \item How many features does our subset have?
  \item Samples?
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Filtering}
  \begin{itemize}
  \item We can make a table to check how many samples we have:
<<eval=TRUE, echo=TRUE>>=
table(ALL_af4bcr$mol.biol)
@  
  \item Our groups are rather different in size, so the outliers of BCR/ABL will dominate the variance.
  \item There are several options on how to filter the data, but we'll use the 10\% and 90\% quantiles.
  \item How do you find that range?
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Filtering II}
  \begin{itemize}
  \item Lets take advantage of the \BIOCfunction{quantile} and \BIOCfunction{diff} functions:
<<eval=TRUE, echo=TRUE>>=
qrange <- function(x)
	diff(quantile(x, c(0.1, 0.9)))
@  
  \item Now we can use the \pl{nsFilter} function from the \pl{genefilter} package:
<<eval=TRUE, echo=TRUE>>=
suppressWarnings(library("genefilter"))
library("hgu95av2.db")
filt_af4bcr <- nsFilter(ALL_af4bcr, require.entrez = TRUE, 
	require.GOBP = TRUE, var.fun = qrange, var.cutoff = 0.5)
ALLfilt_af4bcr <- filt_af4bcr$eset	
@
  \item Previously we had used the \pl{IQR} function instead of our homemade \pl{qrange}.
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Top 100}
  \begin{itemize}
  \item Now lets find the top 100 genes by carrying out a two group comparison.
  \item We'll need to load some packages first:
<<eval=TRUE, echo=TRUE>>=
library("Biobase")
library("annotate")
@  
  \item Now we can use the \BIOCfunction{rowttests} function:
<<eval=TRUE, echo=TRUE>>=
rt <- rowttests(ALLfilt_af4bcr, "mol.biol")
names(rt)
@  
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks]
  \frametitle{Quick exercises}
  Create a histogram of
  \begin{itemize}
  \item the statistic
  \item the p values
  \end{itemize}
\end{frame}

\begin{frame}[fragile, allowframebreaks]
  \frametitle{Solution I}
<<echo=TRUE,fig=TRUE,height=4>>=
hist(rt$statistic, breaks=100, col="skyblue")
@ 
\end{frame}

\begin{frame}[fragile, allowframebreaks]
  \frametitle{Solution II}
<<echo=TRUE,fig=TRUE,height=4>>=
hist(rt$p.value, breaks=100, col="mistyrose")
@ 
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Lowest 400 p values}
  \begin{itemize}
  \item Lets create the \pl{ALLsub} \emph{ExpressionSet} with the 400 probe sets with the lowest p values.
  \item Any ideas?
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Solution}
  \begin{itemize}
  \item Here is one way:
<<eval=TRUE, echo=TRUE>>=
sel <- order(rt$p.value)[1:400]
ALLsub <- ALLfilt_af4bcr[sel,]
@  
  \item Next, lets find how many probe sets in ALL and how many in ALLsub map to the same EntrezGene ID.
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{A trick}
  \begin{itemize}
  \item First lets get the IDs into two separate vectors:
<<eval=TRUE, echo=TRUE>>=
EG <- as.character(hgu95av2ENTREZID[featureNames(ALL)])
EGsub <- as.character(hgu95av2ENTREZID[featureNames(ALLsub)])
@  
  \item Next, lets use a little trick: using two table functions!
<<eval=TRUE, echo=TRUE>>=
head(table(EG))
table(table(EG))
table(table(EGsub))
@  
  \item Why do all the probe sets in ALLsub map to a unique EntrezGene ID?
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Looking at a gene}
  \begin{itemize}
  \item Now lets look at the expression profile of a given gene, for example, CD44.
  \item First, lets find out which features belong to our gene:
<<eval=TRUE, echo=TRUE>>=
syms <- as.character(hgu95av2SYMBOL[featureNames(ALLsub)])
whFeat <- names(which(syms == "CD44"))
@  
  \item Now lets create a subset of ALLsub with the info we want:
<<eval=TRUE, echo=TRUE>>=
ordSamp <- order(ALLsub$mol.biol)
CD44 <- ALLsub[whFeat, ordSamp]
@  
  \item What kind of plot should we make to visualize the expression profile of CD44?
  \end{itemize}
\end{frame}

\begin{frame}[fragile, allowframebreaks]
  \frametitle{Simple plot}
  A simple plot is enough:
<<echo=TRUE,fig=TRUE,height=5>>=
plot(as.vector(exprs(CD44)), main=whFeat, col=c("sienna", "tomato")[CD44$mol.biol], pch=c(15, 16)[CD44$mol.biol], ylab="expression")
@ 
\end{frame}

\begin{frame}[fragile, allowframebreaks]
  \frametitle{Now a barplot}
  We used some mapping tricks to distinguis the two molecular types. \\ Looks like ALL1/AF4 have higher values than BCR/ABL. \\ Now lets make a barplot to group the values per chromosome:
<<echo=TRUE,fig=TRUE,height=5>>=
z <- toTable(hgu95av2CHR[featureNames(ALLsub)])
chrtab <- table(z$chromosome)
chrtab
chridx <- sub("X", "23", names(chrtab))
chridx <- sub("Y", "24", chridx)
barplot(chrtab[order(as.integer(chridx))], cex.names=0.5, col=rainbow(24))
@ 
\end{frame}

\begin{frame}[allowframebreaks]
  \frametitle{Checking}
  \begin{itemize}
  \item Why did I use the \pl{sub} commands?
  \item Why did I use \pl{order} inside the \pl{barplot} call?
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{A sweet html table}
  \begin{itemize}
  \item Now lets assume that you want to show a table for the 400 genes in ALLsub to someone.
  \item Lets use the \pl{annaffy} package to create an html table:
<<eval=TRUE, echo=TRUE>>=
library("annaffy")
anncols <- aaf.handler(chip="hgu95av2.db")[c(1:3, 8:9, 11:13)]
anntable <- aafTableAnn(featureNames(ALLsub), 
	"hgu95av2.db", anncols)
saveHTML(anntable, "ALLsub.html", title="The Features in ALLsub")
@  
  \item We can open the html file directly from R using:
<<eval=FALSE, echo=TRUE>>=
localURL = file.path(getwd(), "ALLsub.html")
browseURL(localURL)
@
  \item Open the html file :)
  \end{itemize}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Complications}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Multiple measurements}
  \begin{itemize}
  \item A big problem is that multiple probe sets can match to the same gene, which means that for some you have more measurements than for others. Also, alternative splicing can give you headaches.
  \item These R packages follow the ENCODE Project Consortium.
  \item Lets look at an example:
<<echo=TRUE, echo=TRUE>>=
probeSetsPerGene <- split(names(EG), EG)
j <- probeSetsPerGene$"7013"
j
@   
  \item We found 7 probes matching to the same gene (EntrezGene ID 7013).
  \end{itemize}
\end{frame}


\begin{frame}[fragile, allowframebreaks]
  \frametitle{Example complication}
  Lets look at the expression values from 2 of them:
<<echo=TRUE,fig=TRUE,height=5>>=
plot(t(exprs(ALL_af4bcr)[j[c(1,7)], ]), asp=1, pch=16,
	col=ifelse(ALL_af4bcr$mol.biol == "ALL1/AF4", "black", "grey"))
@ 
\end{frame}

\begin{frame}[fragile, allowframebreaks]
  \frametitle{A complicated plot}
  We now used a different trick to map the colors: the \BIOCfunction{ifelse} function. \\
  A better plot in this case is the heatmap using the \pl{lattice} function \BIOCfunction{levelplot}. Lets make one for the our gene 7013.
<<echo=TRUE,fig=TRUE,height=5.5>>=
library("lattice")
mat <- exprs(ALL_af4bcr)[j,]
mat <- mat - rowMedians(mat)
ro <- order.dendrogram(as.dendrogram(hclust(dist(mat))))
co <- order.dendrogram(as.dendrogram(hclust(dist(t(mat)))))
at <- seq(-1, 1, length=21) * max(abs(mat))
lp <- levelplot(t(mat[ro, co]),
	aspect = "fill", at = at,
	scales = list(x = list(rot=90)),
	colorkey = list(space = "left"))
print(lp)
@ 
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Some statistics}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{chr}
  \begin{itemize}
  \item One of  the tests we can make now is to check for every chromosome, the low and high p values.
  \item To do so we can use \pl{chisq.test} and \pl{fisher.test}.
  \item First we need to create a data frame to map for every EntrezGene ID to which chromosome it belongs:
<<eval=TRUE, echo=TRUE>>=
ps_chr <- toTable(hgu95av2CHR)
ps_eg <- toTable(hgu95av2ENTREZID)
chr <- merge(ps_chr, ps_eg)
dim(chr)
@  
  \item We don't need the first column, so lets take it out:
<<eval=TRUE, echo=TRUE>>=
chr <- unique(chr[, colnames(chr)!="probe_id"])
dim(chr)
head(chr)
@  
  \item What problem do you notice? You might need to explore \pl{chr} in full.
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Duplications}
  \begin{itemize}
  \item Look at this table:
<<eval=TRUE, echo=TRUE>>=
table(table(chr$gene_id))
@  
  \item Lets take out those complicated genes that have duplicated entries.
<<eval=TRUE, echo=TRUE>>=
chr <- chr[!duplicated(chr$gene_id), ]
@  
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Checking for association}
  \begin{itemize}
  \item Now we can do the a contigency table for the association between the EntrezGene ID with their chromosome mapping \alert{and} with being differently expressed.
  \item Lets re-use our EGsub object which had those differently expressed.
<<eval=TRUE, echo=TRUE>>=
isdiff <- chr$gene_id %in% EGsub
tab <- table(isdiff, chr$chromosome)
tab
@  
  \item Once we have this table, we can do a Fisher's exact test:
<<eval=TRUE, echo=TRUE>>=
fisher.test(tab, simulate.p.value=TRUE)
@
  \item And a Chi squared test:
<<eval=TRUE, echo=TRUE>>=
chisq.test(tab)
@  
  \item What can we conclude?
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Strand bias}
  \begin{itemize}
  \item We can also check for where the genes are located, what other genes are nearby, grouping genes by location before another test, \ldots
  \item Lets check if our differentially expressed genes are on the same strand:
<<eval=TRUE, echo=TRUE>>=
chrloc <- toTable(hgu95av2CHRLOC[featureNames(ALLsub)])
head(chrloc)
@  
  \item Alternative splicing will give us some problems:
<<eval=TRUE, echo=TRUE>>=
table(table(chrloc$probe_id))
@
  \item Lets collapse the information so that we only record the strand, which should be the same even if there is alternative splicing:
<<eval=TRUE, echo=TRUE>>=
strds <- with(chrloc, unique(cbind(probe_id, sign(start_location))))
table(strds[,2])
@  
  \item What do we conclude?
  \end{itemize}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{GO}

\begin{frame}[allowframebreaks]
  \frametitle{Quick review}
  \begin{itemize}
  \item GO, short for Gene Ontology, classifies genes products according to
  \begin{enumerate}
  \item Molecular function
  \item Biological process
  \item Cellular component
  \end{enumerate}
  \item GO terms are represented in a graph where there are two types of relationships:
  \begin{enumerate}
  \item is as
  \item part of
  \end{enumerate}
  \item To facilitate the mapping, GO terms are identified in 7 numbers. 
  \item All the descendants of a given GO term are called \emph{offspring}. The immediate ones are called \emph{children}.
  \item All the parental GO terms are called \emph{ancestor}.
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{GO.db}
  \begin{itemize}
  \item In R, the package \pl{GO.db} enables us to browse the GO tree:
<<eval=TRUE, echo=TRUE>>=
library("GO.db")
as.list(GOMFCHILDREN["GO:0008094"])
as.list(GOMFOFFSPRING["GO:0008094"])
@  
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Hyper Geometric GO test}
  \begin{itemize}
  \item The packages \alert{annotate} and \alert{GOstats} are the basic ones to carry out GO analysis.
  \item Other related packages are \BIOCfunction{topGO} and \BIOCfunction{goTools}.
  \item Lets make the basic GO test. We want to compare the frequency of a GO term on a subset versus the frequency of the same GO term on the overall universe.
  \item Things get complicated because some GO terms have more offspring than others\ldots
  \item Lets do the test (actually, lots of tests) for our data:
<<eval=TRUE, echo=TRUE>>=
library("GOstats")
affyUniverse <- featureNames(ALLfilt_af4bcr)
uniId <- hgu95av2ENTREZID[affyUniverse]
entrezUniverse <- unique(as.character(uniId))
params <- new("GOHyperGParams", 
	geneIds = EGsub, universeGeneIds = entrezUniverse,
	annotation = "hgu95av2", ontology = "BP",
	pvalueCutoff =0.001, conditional = FALSE,
	testDirection = "over")
@  
  \item After building up all the parameters we can now make the actual test:
<<eval=TRUE, echo=TRUE>>=
myhyper <- hyperGTest(params)
@  
  \end{itemize}
\end{frame}

\begin{frame}[fragile, allowframebreaks]
  \frametitle{P values}
  We didn't adjust our p values as it can complicated. Instead, lets visualize the histogram:
<<echo=TRUE,fig=TRUE,height=5>>=
hist(pvalues(myhyper), breaks=50, col="mistyrose")
@ 
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Summary for myhyper}
  \begin{itemize}
  \item As you can notice, we have a peak on the left side. Meaning that we have several low p values.
  \item Lets look deeper into the results from our test: \scriptsize
<<eval=TRUE, echo=TRUE>>=
sum <- summary(myhyper, p=0.001)
head(sum)
@  
  \normalsize
  \item What do you notice? What can you conclude?
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{Longer definitions}
  \begin{itemize}
  \item Even though the GO term definition is better than the GO ID, it is not sufficient.
  \item So lets take a look at the actual definitions using the \pl{GO.db} package:
<<eval=TRUE, echo=TRUE>>=
GOTERM[["GO:0032945"]]
@  
  \end{itemize}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{More on GO}

\begin{frame}[allowframebreaks]
  \frametitle{biomaRt}
  \begin{itemize}
  \item Remember that you can use \BIOCfunction{biomaRt} to get GO IDs or to use them as a query and get more information on your genes / proteins.
  \item For instance, take a look at the \BIOCfunction{getGo} function.
  \item You can find GO IDs from \pl{biomaRt} in PFAM, Prosite, and InterPro besides the usual, ENSEMBL.
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks]
  \frametitle{SQL based packages}
  \begin{itemize}
  \item Several packages, for example hgu133a and hgu95av2 were changed from being \emph{environment} based to SQL based packages.
  \item They did this change to facilitate mapping between different identifiers.
  \item This was specially useful in cases where you have incomplete data.
  \item Plus it made everything faster :)
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{An example:}
  \begin{itemize}
  \item Old way:
<<eval=TRUE, echo=TRUE>>=
goCats <- unlist(eapply(GOTERM, Ontology))
old <- table(goCats)[c("BP", "CC", "MF")]
@
  \item New way \alert{WAY faster}:  
<<eval=TRUE, echo=TRUE>>=
query <- "select ontology from go_term"
goCats <- dbGetQuery(GO_dbconn(), query)
new <- table(goCats)[c("BP", "CC", "MF")]
@  
  \item Comparing:
<<eval=TRUE, echo=TRUE>>=
identical(old, new)
@  
  \end{itemize}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Credits}

\begin{frame}[allowframebreaks]
  \frametitle{Credits}
  \begin{itemize}
  \item \BIOCfunction{Bioconductor Case Studies} by Florian Hahne, Wolfgang Huber, Robert Gentleman and Seth Falcon.
  \item Specially chapter 8.
  \end{itemize}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Homework}
\begin{frame}[allowframebreaks]
  \frametitle{Homework}
  \begin{itemize}
  \item Choose a different EntrezGene ID (not 7013) that has different probes.
  \item Make a scatterplot compairing the expression values from two probe sets.
  \item Make the heatmap showing all the probe sets.
  \item Add your conclusions.
  \end{itemize}
\end{frame}

\begin{frame}[allowframebreaks, fragile]
  \frametitle{SessionInfo} \scriptsize
<<eval=TRUE, echo=TRUE>>=
sessionInfo()
@  
\end{frame}

\end{document}

